### PR TITLE
Issue 5659 - local fonts on editor

### LINKED ISSUE #5659: Load local fonts on editor
### Description

Backend still loads the fonts from the external bunny site, if local option is enabled, is must be served from the local folder on backend too. (same for google fonts)

_Originally posted by @elzadj in https://github.com/maxi-blocks/maxi-blocks/pull/5642#pullrequestreview-2153570604_

### Step-by-step reproduction instructions

1. Open dashboard settings, enable local fonts.
2. On the editor add any font to text maxi and in the console check where the font is loaded from.

### Screenshots, screen recording, code snippet

image
image

### URL and/or Code editor

_No response_

### How to fix it

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by myroslavve**
# Description
This update makes loading fonts slower for local fonts, because we have to wait for the font to load from google/bunny to local server and then wait for fonts to load from server to editor. There is nothing we can do about it, maybe we should add spinning wheel to the font selector to show when the font is loading now.

Fixes #5659 

# How Has This Been Tested?
Enabled local fonts in the dashboard, on editor loaded any font, checked in console that path to that font is "localhost...".

# Test checklist

**_ Front/Back Testing _**

-   [ ] Test the component in normal state in sidebar
-   [ ] Test the component in hover state in sidebar (if hover exists)
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points for the toolbar
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type

**_ Pre-Code Testing _**

-   [ ] Test the component in normal state in sidebar
-   [ ] Test the component in hover state in sidebar (if hover exists)
-   [ ] Check that the settings work on frontend
-   [ ] Check that backend works and saves the settings after the editor reload
-   [ ] Same 1-4 points for the toolbar
-   [ ] Same 1-4 points on responsive
-   [ ] Test the block inside the grid (Container + Row + Column)
-   [ ] Test the block as a standalone block
-   [ ] Duplicate the block, test that the settings of the first do not affect the second
-   [ ] Test with 2 blocks of the same type
-   [ ] Check no commented code and no unnecessary imports
-   [ ] Standards of the project have been followed
-   [ ] No errors/warnings on console

# Checklist

-   [ ] My code follows the style guidelines of this project
-   [ ] I have performed a self-review of my code
-   [ ] I have commented my code, particularly in hard-to-understand areas
-   [ ] My changes generate no new warnings/errors
-   [ ] I have added/updated tests that prove my fix is effective or that my feature works
-   [ ] New and existing unit tests pass locally with my changes

## Summary by CodeRabbit

- **New Features**
  - Introduced a new REST API endpoint to dynamically retrieve font URLs.
  - Enhanced loading state management across various components for a better user experience during font operations.
  - Added an `overlay` prop to the ContentLoader component for dynamic styling capabilities.

- **Bug Fixes**
  - Improved error handling and clarity in font CSS file uploads.

- **Refactor**
  - Simplified font URL generation logic for improved maintainability.
  - Modularized font URL construction and loading logic for better readability and clarity.
  - Transitioned to a singleton pattern for font management.
  - Improved naming conventions for methods and properties to enhance code clarity.

### CODE REVIEW

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — CHANGES_REQUESTED** [review]
Getting weird results after I re-open the editor:

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.07.26-15_32_59 (1).webm](https://github.com/user-attachments/assets/41a82b5c-b403-426e-8e92-3bed6f762d2a)

Tests tell us something is breaking, too.

Great idea about a loader, but I would add it to the Text (Button, etc) block itself, maybe with "Loading font..." text, but I guess the one near the Font selector is easier to do.

**Comment by elzadj** [comment]
@Fybex test this one too please, when you have time. Thanks!

**Comment by elzadj** [comment]
@myroslavve @Olekrut can't recreate the bugs from my video, btw, so could be something local.
But the tests are failing hard, so something is not right, needs checking.

File: src/extensions/text/store/resolvers.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Wordpress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+
+const resolvers = {
+	getFontUrl:
+		(fontName, fontData) =>
+		async ({ dispatch }) => {
+			const fontUrl = JSON.parse(
```
Comment (Fybex) [inline src/extensions/text/store/resolvers.js:10]: Looks like JSON.parse is unnecessary if avoid double encoding by removing wrapping url in [class-maxi-api](https://github.com/maxi-blocks/maxi-blocks/blob/b66b98a04a5a02e301fb8ece48e692239f372747/core/class-maxi-api.php#L704)

Comment (myroslavve) [reply src/extensions/text/store/resolvers.js]: sure, this works :+1:

**Comment by myroslavve** [comment]
@Olekrut test with bunny fonts again pls

**Review by Olekrut — APPROVED** [review]
image

**Review by elzadj — CHANGES_REQUESTED** [review]
Almost, almost, great work so far! 

1. I still see a space in the id on backend if the font name has 2 and more spaces. Need to replace globally.
Screenshot from 2024-08-02 16-22-49

2. There is a several seconds visual delay when we change a font while it's uploading, we really need some sort of indicator here. @ckp267 @kyrapieterse ? I think we should add a loading animation to the whole text blocks, maybe even re-use our usual loader, but with some semi-opacity for the block.

[screencast-wordpress.local-2024.08.02-16_21_14.webm](https://github.com/user-attachments/assets/366a0e6b-4c02-4d9f-9450-4c9d551bbaf8)

**Comment by ckp267** [comment]
> 2. There is a several seconds visual delay when we change a font while it's uploading, we really need some sort of indicator here. @ckp267 @kyrapieterse ? I think we should add a loading animation to the whole text blocks, maybe even re-use our usual loader, but with some semi-opacity for the block.
> 
>  screencast-wordpress.local-2024.08.02-16_21_14.webm

As per @elzadj 

**My idea for the loading - make the text 0.5 opacity and show our usual loader circle.**

---
Note from Llama 405B as a summary of internal discussion thread:
- The delay in displaying a font is due to the need to download it first. To address this, a loading indicator will be implemented, which will display a loader circle and reduce the text opacity to 0.5. 
- The idea of previewing the font style in the dropdown was discussed, but it was deemed impractical due to the need to load all fonts or create images for each font. 
- Instead, the focus will be on implementing a loader that indicates the font is being downloaded. Unused fonts will be removed periodically, specifically each time all fonts are loaded, to prevent accumulation of unnecessary files. 
- The loader will not include any text, such as "Downloading fonts...", due to potential layout and language translation issues.

**Review by elzadj — CHANGES_REQUESTED** [review]
Works great! Now need to add it to:
1. Image caption text font
2. Number Counter number
3. Accordion 
4. Search input


@Olekrut did I miss something?

**Review by Olekrut — CHANGES_REQUESTED** [review]
Then need to do the same for SC as well
and for Map block

About loader:
Looks like with loader is longer than on master
this branch:
https://jam.dev/c/3329405e-b3c9-4c53-aeb5-80fe453dded5
master:
https://jam.dev/c/57e79f36-9863-44c0-8b0d-d867e5aded8c

**Comment by elzadj** [comment]
@myroslavve great! 
Toolbar needs it too:
screenshot-wordpress local-2024 08 09-12_56_35

And SC, but not sure where to add the loader here - main thing for it to be visible.

**Comment by myroslavve** [comment]
> Then need to do the same for SC as well and for Map block

Added.

> About loader: Looks like with loader is longer than on master this branch: https://jam.dev/c/3329405e-b3c9-4c53-aeb5-80fe453dded5 master: https://jam.dev/c/57e79f36-9863-44c0-8b0d-d867e5aded8c

The reason we added the loader is that other changes in this PR made loading fonts slower.

**Review by elzadj — CHANGES_REQUESTED** [review]
Niice, found that for H1 in a custom SC the loader gets stuck (and then all other loaders after it, too). I've used an old pattern from the master.
screenshot-wordpress local-2024 08 20-13_01_54

**Comment by myroslavve** [comment]
> Niice, found that for H1 in a custom SC the loader gets stuck (and then all other loaders after it, too). I've used an old pattern from the master. screenshot-wordpress local-2024 08 20-13_01_54

Do you have any errors in console? I fixed this issue yesterday and cannot reproduce it on the latest version of this branch.

**Comment by elzadj** [comment]
> Do you have any errors in console? I fixed this issue yesterday and cannot reproduce it on the latest version of this branch.

Not really:

[screencast-bpconcjcammlapcogcnnelfmaeghhagj-2024.08.20-17_47_27.webm](https://github.com/user-attachments/assets/7c33a0ef-cdae-4c7e-ae44-8ef8ad116c9c)

Happens both with google and bunny fonts

**Comment by elzadj** [comment]
@Olekrut Do you see the same bug? ^

**Comment by Olekrut** [comment]
> Do you see the same bug? ^

Actually, i can re-create it everywhere ( P, button, H1-6 ) if change font again before it loads.
Everything is gets ok if to wait 2-3 min, but yes, it is not ok

**Comment by elzadj** [comment]
@myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.

**Comment by kyrapieterse** [comment]
> @myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.

Agree = remove what we dont need

**Comment by elzadj** [comment]
> > @myroslavve what's the status of this one? Maybe let's remove the loader from the SC for now, if it's easier to do? Then we can check if we have any support requests about it.
> 
> I still cannot consistently reproduce this issue, so let's do it separately, because it can take some time to figure out.

Good idea, but remove the loader from the SC for now, please, in case we need to update the plugin before it's fixed. Thanks!

**Comment by elzadj** [comment]
I merge it as soon as the tests pass. @myroslavve create a new issue for the SC loader, please.

**Review by elzadj — CHANGES_REQUESTED** [review]
Something's wrong with frontend when I use bunny local

[screencast-wordpress.local-2024.08.27-18_00_39 (1).webm](https://github.com/user-attachments/assets/42ced772-bff8-48b3-ab17-6d032f864010)

**Review by Olekrut — CHANGES_REQUESTED** [review]
the same with google fonts, but works only custom  fonts and only on frontend
image

**Comment by elzadj** [comment]
@myroslavve not sure how it can affect the backgrounds, but the tests clearly show it :melting_face:

**Review by Olekrut — CHANGES_REQUESTED** [review]
image
frontend

**Comment by myroslavve** [comment]
> image frontend

fixed

**Review by Olekrut — CHANGES_REQUESTED** [review]
https://jam.dev/c/df9fc381-e949-455a-bcdb-2b9849729aa1
local fonts are loading very long (3-4min) on backend
https://jam.dev/c/f03c5d85-491f-40e4-b108-f7ca523fb308

**Comment by myroslavve** [comment]
> https://jam.dev/c/df9fc381-e949-455a-bcdb-2b9849729aa1 local fonts are loading very long (3-4min) on backend https://jam.dev/c/f03c5d85-491f-40e4-b108-f7ca523fb308

Tried to optimize, but everything is already as fast as it can be for local fonts, at least logically. With local fonts the speed depends a lot on the speed of the server, so on local it will be slower than on an actual server where WordPress would be.

**Review by Olekrut — APPROVED** [review]

**Review by elzadj — APPROVED** [review]

**Pull request merged**